src/gravity_hyperbolic.f90

```diff
@@ -102,28 +103,33 @@ subroutine hyperbolic_gravity_step(cgrav_now,cgrav_old,dtg)
 !$omp end do
 
 ! Smear gravity in central regions -----------------------------------
+!$omp single
   if(fmr_max>0)then
    do n = 1, fmr_max
     if(fmr_lvl(n)==0)cycle
-    jb=min(2**(fmr_max-n+1),je)-1 ; kb=min(2**(fmr_max-n+1),ke)-1
+    jb=min(2**(fmr_max-n+1),je_global)-1 ; kb=min(2**(fmr_max-n+1),ke_global)-1
     if(n==1)then
-     jb=je-js; kb=ke-ks
+     jb=je_global-js_global; kb=ke_global-ks_global
     end if
-!$omp do private(i,j,k,vol) collapse(3)
-    do k = ks, ke, kb+1
-     do j = js, je, jb+1
-      do i = is+sum(fmr_lvl(0:n-1)), is+sum(fmr_lvl(0:n))-1
+    do k = ks_global, ke_global, kb+1
+     do j = js_global, je_global, jb+1
+      do i = is_global+sum(fmr_lvl(0:n-1)), is_global+sum(fmr_lvl(0:n))-1
        vol = sum(dvol(i,j:j+jb,k:k+kb))
-       grvphi(i,j:j+jb,k:k+kb) = sum(grvphi(i,j:j+jb,k:k+kb) &
-                                      *dvol(i,j:j+jb,k:k+kb)) / vol
-       grvpsi(i,j:j+jb,k:k+kb) = sum(grvpsi(i,j:j+jb,k:k+kb) &
-                                      *dvol(i,j:j+jb,k:k+kb)) / vol
+
+       il = max(i,is); ir = min(i,ie)
+       jl = max(j,js); jr = min(j+jb,je)
+       kl = max(k,ks); kr = min(k+kb,ke)
+
+       grvphi(il:ir,jl:jr,kl:kr) = &
```

Comment:
Are we relying here that if `il>ir` etc, the array is an empty array so no operations will be done?

Reply:
Yes, well spotted. I've added a code comment here for future reference because I don't think it's immediately clear that it does this.